File: Gemini/Intermediárias/gale_shapley/Gale-Shapley.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Student {
    int id;
    vector<int> preferredHospitals;
};

struct Hospital {
    int id;
    vector<int> preferredStudents;
};

int main() {
    int numStudents, numHospitals;
    cin >> numStudents >> numHospitals;

    vector<Student> students(numStudents);
    vector<Hospital> hospitals(numHospitals);

    // Read student and hospital preferences
    for (int i = 0; i < numStudents; ++i) {
        students[i].id = i;
        // Read preferred hospitals for student i
    }

    for (int i = 0; i < numHospitals; ++i) {
        hospitals[i].id = i;
        // Read preferred students for hospital i
    }

    // Gale-Shapley algorithm
    while (true) {
        bool foundMatch = false;
        for (Student& student : students) {
            if (student.preferredHospitals.empty()) {
                continue;
            }

            int hospitalId = student.preferredHospitals.front();
            student.preferredHospitals.erase(student.preferredHospitals.begin());

            if (hospitals[hospitalId].preferredStudents.empty() || hospitals[hospitalId].preferredStudents.front() > student.id) {
                hospitals[hospitalId].preferredStudents.clear();
                hospitals[hospitalId].preferredStudents.push_back(student.id);
                foundMatch = true;
            }
        }

        if (!foundMatch) {
            break;
        }
    }

    // Print the stable matching
    for (const Hospital& hospital : hospitals) {
        if (!hospital.preferredStudents.empty()) {
            cout << "Student " << hospital.preferredStudents.front() << " is matched with Hospital " << hospital.id << endl;
        }
    }

    return 0;
}

File: Gemini/Intermediárias/gale_shapley/Gale-Shapley.java
import java.util.*;

class Student {
    String name;
    List<Hospital> preferences;

    public Student(String name, List<Hospital> preferences) {
        this.name = name;
        this.preferences = preferences;
    }
}

class Hospital {
    String name;
    List<Student> preferences;

    public Hospital(String name, List<Student> preferences) {
        this.name = name;
        this.preferences = preferences;
    }
}

public class StableMatching {
    public static Map<Student, Hospital> findStableMatching(List<Student> students, List<Hospital> hospitals) {
        Map<Student, Hospital> matches = new HashMap<>();
        Map<Hospital, Student> hospitalMatches = new HashMap<>();

        // Initialize all hospitals to free
        for (Hospital hospital : hospitals) {
            hospitalMatches.put(hospital, null);
        }

        while (!students.isEmpty()) {
            Student student = students.remove(0);
            Hospital hospital = student.preferences.get(0);

            if (hospitalMatches.get(hospital) == null) {
                matches.put(student, hospital);
                hospitalMatches.put(hospital, student);
            } else {
                Student currentMatch = hospitalMatches.get(hospital);

                if (hospital.preferences.indexOf(student) < hospital.preferences.indexOf(currentMatch)) {
                    matches.put(student, hospital);
                    hospitalMatches.put(hospital, student);
                    students.add(currentMatch);
                }
            }

            student.preferences.remove(0);
        }

        return matches;
    }

    public static void main(String[] args) {
        // ... (create students and hospitals with preferences)
        List<Student> students = ...;
        List<Hospital> hospitals = ...;

        Map<Student, Hospital> stableMatching = findStableMatching(students, hospitals);

        // Print the stable matching
        for (Map.Entry<Student, Hospital> entry : stableMatching.entrySet()) {
            System.out.println(entry.getKey().name + " matched with " + entry.getValue().name);
        }
    }
}